python/activator/activator.py

```diff
@@ -257,73 +324,95 @@ def keda_start():
-
-                # Commit message and close client
-                fan_out_consumer.commit(message=fan_out_message, asynchronous=False)
-                fan_out_consumer.close()
+                try:
+                    fan_out_message = redis_client.xreadgroup(
+                        streams={redis_stream_name: ">"},
+                        consumername=redis_group_id,
+                        groupname=redis_stream_consumer_group,
+                        count=1,)
 
+                    if not fan_out_message:
+                        continue
+                    else:
+                        consumer_polls_with_message += 1
+                        if consumer_polls_with_message >= 1:
+                            fan_out_listen_time = _calculate_time_since_last_message(
+                                fan_out_listen_start_time)
+                            _log.debug(
+                                "Seconds since last redis streams message received %r for consumer poll %r",
+                                fan_out_listen_time, consumer_polls_with_message)
+
+                        redis_streams_message_id, fan_out_visit_decoded = _decode_redis_streams_message(
+                            fan_out_message)
+
+                        # Calculate time to receive message based on timestamp in Redis Stream message
+                        fan_out_to_prompt_time = _calculate_time_since_fan_out_message_delivered(
+                            redis_streams_message_id)
+                        _log.debug("Seconds since fan out message delivered %r", fan_out_to_prompt_time)
+
+                        expected_visit = FannedOutVisit.from_dict(fan_out_visit_decoded)
+                        _log.debug("Unpacked message as %r.", expected_visit)
+
+                        # Ack the redis stream message and close redis stream client
+                        # TODO Consider moving xack after process visit completes for catch up processing.
+                        redis_client.xack(redis_stream_name,
+                                          redis_stream_consumer_group,
+                                          redis_streams_message_id)
+                        redis_client.close()
+                # TODO Review Redis Errors and determine what should be retriable.
+                except redis.exceptions.RedisError as e:
+                    _log.critical("Redis Streams error; aborßting.")
+                    _log.exception(e)
+                    redis_client.close()
+                    sys.exit(1)
+
+                with instances_processing_gauge.track_in_progress():
+                    try:
+                        # Process fan out visit
+                        process_visit(expected_visit)
+                    except Exception as e:
+                        _log.critical("Process visit failed; aborting.")
+                        _log.exception(e)
```

Comment:
A couple of things going on here:
- `critical` should be reserved for fatal errors (i.e., those that can only be handled by killing the worker); errors that only abort the current processing run should be `error`
- `exception` actually gets the exception object directly from the `except` block. The argument is interpreted as a string message, same as any other logging command
- `exception` already logs at `error` level, so it will get picked up by that filter

I propose the following, analogous to `server_error`:
```suggestion
                    except Exception as e:
                        _log.exception("Processing failed:")
```

Reply:
Even if Keda can't do retries yet, some of the other exceptions raised by `process_visit` can be handled (text copied from the Flask code):
```
except activator.exception.GracefulShutdownInterrupt:
    _log.error("Service interrupted. Shutting down *without* syncing to the central repo.")
    sys.exit(1)
except activator.exception.IgnorableVisit as e:
    _log.info("Skipping visit: %s", e)
```

(None of our dev instances are configured to ignore particular visits, so the second clause wouldn't have come up in your tests.)